### DISCUSSION TITLE
Offline Enabled App with SWR

### DISCUSSION THREAD

**Discussion by niicojs** [vercel/swr#696] [Q&A]
Hi guys,

I'm trying to add offline capabilities to a small app I made that use SWR.
What I have in mind is using IndexedDB as a local cache provider.

A solution would be adding cache update and fallback into the fetcher func, but it seems heavy for something that I guess already exists somewhere.
Or maybe populate SWR at startup from DB and update cache in fetcher so I don't have to test is we are online in fetcher.

How are you guys doing this?

Thanks,

**Comment by huozhi** [comment]
@niicojs using `mutate(key, data, shouldRevalidate?)` API might help on this case. when you'd like to pull states from DB and patch them to swr cache, `mutate` is definitely for that.

```js
import {mutate} from 'swr'

// init DB
db.connect()
// ...
// patch states to cache
const state = await db.query(.....)
mutate(key, state, false)
```

swr might try to bring customized cache in the future (I guess?), any thoughts or requirements are welcomed to post here.

**Reply by niicojs** [reply]
For now, I've called the cache API directly, but you're right as it's not really documented I should maybe use the mutate api.

**Reply by huozhi** [reply]
Cache API is in the early and unstable phase, might change in the future. Once it’s settled SWR will document it in details for sure.
